### DISCUSSION TITLE
[FE210] 關於Twitch API 串接

### DISCUSSION THREAD

**Discussion by dracarys55** [Lidemy/forum#138] [前端相關問題]
內文：

作業名稱:
Ajax 與 API 串接

碰到的問題: 
按照老師寫的程式碼進行串接後,卻遇到了  CORS policy 同源政策的錯誤

預期的行為:
取得Twitch API 的資料,以JSON呈現

嘗試過的解決方法:
1.遇到問題之後,先按照其他網路文章串接了Unsplash 的API 可以正常運作
2.爬了之前在Forum 唯一一篇的[FE210]  的文章並複製了那位同學的程式碼先進行使用 但是一樣有問題 但是無能力進行修改XD
3.加入了一條變數宣告 想要試試看是不是要發Token給Twitch API 但是顯示一樣的  CORS policy error
4.最後來到論壇求助

相關程式碼:
const clientId = "dzafgpctd55n19kzgyh83gjsi8b95s";
const appToken = "k70qjfu6yzqlmk3dg7pkcyvhf34cbw";
const api =
  "https://api.twitch.tv/kraken/streams/?game=League%20of%20Legends&limit=2";

let xhr = new XMLHttpRequest();
xhr.open("GET", api, true);
xhr.setRequestHeader("client-id", clientId);
xhr.setRequestHeader("Authorization", appToken);
xhr.send();

xhr.onreadystatechange = function () {
  if ((this.readyState === 4) & (this.status === 200)) {
    var data = JOSN.parse(this.responseText);
    console.log(data);
  }
};

補充說明:
懷疑應該是Twitch 有進行更新 必須要再給更多的權杖認證?
已試過goole上面的各式各樣文章的方法 
無奈某些沒有結果
有些是自己沒有能力看懂
只好上來求助各位同學老師了 感恩

**Comment by allenliaooooooo** [comment]
Twitch 官方 API 文件說要加 accept header 有加嗎？

**Reply by outshaker** [reply]
幫樓上補充更詳細一點，正確來說，需要加 accept header 的是舊版本 v5 API，然後我看同學的程式碼有出現 app token，感覺上有被官方新版本的 API 文件影響。

**Comment by outshaker** [comment]
- 舊版 API (官方文件名稱 Legacy Twitch API v5)
  - 參考文件是這個 https://dev.twitch.tv/docs/v5 代號 kraken
  - header 需要設定 Accept 和 Client-ID

- 新版 API (官方文件名稱 Twitch API)
  - 參考文件是這個 https://dev.twitch.tv/docs/api 代號 helix
  - header 需要設定 Authorization 和 Client-ID

建議同學以官方文件寫的內容為主，因為它的資訊和內容是最正確的，因為別人的寫法很有可能寫錯或者混淆。誠心建議就算要抄 code 也先抄官方文件給的範例 :)

以下是針對你的程式碼給的一些小回饋：

```js
    function call() {
      const clientId = "dzafgpctd55n19kzgyh83gjsi8b95s"
      const appToken = "k70qjfu6yzqlmk3dg7pkcyvhf34cbw" // 這是新版才需要的東西
      const api = "https://api.twitch.tv/kraken/streams/?game=League%20of%20Legends&limit=2" // kraken 是舊版的代號，所以下面要按照舊版的規定走

      let xhr = new XMLHttpRequest()
      xhr.open("GET", api, true) // async = true 是預設設定，其實不用多打這個
      xhr.setRequestHeader('Accept', 'application/vnd.twitchtv.v5+json') //舊版要加這個
      xhr.setRequestHeader("client-id", clientId)
      xhr.setRequestHeader("Authorization", appToken) // 舊版不需要這個
      xhr.send()

      xhr.onreadystatechange = function () {
        if ((this.readyState === 4) & (this.status === 200)) {
          var data = JOSN.parse(this.responseText) // JSON 打錯了囉
          console.log(data)
        }
      }
    }
```

**Reply by dracarys55** [reply]
非常感謝解答!!  終於有看到資料了
謝謝同學耐心的解答 非常受用
下次一定會慢慢去看官方API文件的
受用無窮 謝謝了

**Reply by outshaker** [reply]
不客氣^^
